pkg/model/awsmodel/oidc_provider.go

```diff
@@ -43,34 +64,156 @@ const (
-	if issuerURL == "" {
-		return nil
+	discovery, err := buildDicoveryJSON(serviceAccountIssuer)
+	if err != nil {
+		return err
+	}
+	keysFile := &fitasks.ManagedFile{
+		Contents:  keys,
+		Lifecycle: b.Lifecycle,
+		Location:  fi.String("oidc/keys.json"),
+		Name:      fi.String("keys.json"),
+		Base:      fi.String(b.Cluster.Spec.PublicDataStore),
 	}
+	c.AddTask(keysFile)
+
+	discoveryFile := &fitasks.ManagedFile{
+		Contents:  fi.NewBytesResource(discovery),
+		Lifecycle: b.Lifecycle,
+		Location:  fi.String("oidc/.well-known/openid-configuration"),
+		Name:      fi.String("discovery.json"),
+		Base:      fi.String(b.Cluster.Spec.PublicDataStore),
+	}
+	c.AddTask(discoveryFile)
 
 	c.AddTask(&awstasks.IAMOIDCProvider{
 		Name:        fi.String(b.ClusterName()),
 		Lifecycle:   b.Lifecycle,
-		URL:         fi.String(issuerURL),
+		URL:         fi.String(serviceAccountIssuer),
 		ClientIDs:   []*string{fi.String(defaultAudience)},
 		Thumbprints: thumbprints,
 	})
 
 	return nil
 }
+
+func buildDicoveryJSON(issuerURL string) ([]byte, error) {
+	d := oidcDiscovery{
+		Issuer:                fmt.Sprintf("%v/", issuerURL),
+		JWKSURI:               fmt.Sprintf("%v/keys.json", issuerURL),
+		AuthorizationEndpoint: "urn:kubernetes:programmatic_authorization",
+		ResponseTypes:         []string{"id_token"},
+		SubjectTypes:          []string{"public"},
+		SigningAlgs:           []string{"RS256"},
+		ClaimsSupported:       []string{"sub", "iss"},
+	}
+	return json.MarshalIndent(d, "", "")
+}
+
+type KeyResponse struct {
+	Keys []jose.JSONWebKey `json:"keys"`
+}
+
+type OIDCKeys struct {
+	SigningKey *fitasks.Keypair
+}
+
+// GetDependencies adds CA to the list of dependencies
+func (o *OIDCKeys) GetDependencies(tasks map[string]fi.Task) []fi.Task {
+	return []fi.Task{
+		o.SigningKey,
+	}
+}
+func (o *OIDCKeys) Open() (io.Reader, error) {
+
+	certBytes, err := fi.ResourceAsBytes(o.SigningKey.Certificate())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cert: %w", err)
+	}
+	block, _ := pem.Decode(certBytes)
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse cert: %w", err)
+	}
+
+	publicKey := cert.PublicKey
+
+	publicKeyDERBytes, err := x509.MarshalPKIXPublicKey(publicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize public key to DER format: %v", err)
+	}
+
+	hasher := crypto.SHA256.New()
+	hasher.Write(publicKeyDERBytes)
+	publicKeyDERHash := hasher.Sum(nil)
+
+	keyID := base64.RawURLEncoding.EncodeToString(publicKeyDERHash)
+
+	keys := []jose.JSONWebKey{
+		{
+			Key:       publicKey,
+			KeyID:     keyID,
+			Algorithm: string(jose.RS256),
+			Use:       "sig",
+		},
+	}
+
+	keyResponse := KeyResponse{Keys: keys}
+	jsonBytes, err := json.MarshalIndent(keyResponse, "", "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal json: %w", err)
+	}
+
+	return bytes.NewReader(jsonBytes), nil
+}
+
+func getFingerprint(issuer string) (string, error) {
```

Comment:
This one will be interesting as S3 is changing certs these days. Luckily we should be good as this most likely won't make it in until 23rd March :)

Reply:
I think it might be better to either hardcode the CA cert fingerprints and/or allow them to be specified in the cluster spec. Getting them by observing what you get from a connection allows an MITM attacker to substitute out the last-listed cert.

Reply:
My reasoning behind this was if someone wanted a different store than S3. But for well-known ones, I suppose hardcoding the fingerprints makes sense.

Reply:
For stores other than S3 (or well-known ones) we could support a field in the cluster spec.

Reply:
True. I think for the time being, only s3 will be supported anyway. So may as well add those two fingerprints.